torch/csrc/autograd/engine.cpp

```diff
@@ -225,21 +225,37 @@ Engine::Engine() : max_recursion_depth_(MAX_DEPTH), non_reentrant_device_thread_
 // Send shutdown tasks to all device_ready_queues_ if no backward tasks are running
 // Even though readyQueue should be empty, shutdown tasks have the highest priority
 Engine::~Engine() {
+  // Occasionally autograd threads can hang on shutdown
+  // Do not wait for them to shutdown indefinitely but rely on timeout
+  auto wait_duration_str = getenv("TORCH_AUTOGRAD_SHUTDOWN_WAIT_LIMIT");
+  if (!wait_duration_str) {
+    wait_duration_str = "10.0";
+  }
+  using namespace std::chrono_literals;
+  auto wait_duration = std::atof(wait_duration_str) * 1.0s;
   bool noBackward = true;
   for (auto& queue: device_ready_queues_) {
     noBackward =  noBackward && queue->empty();
   }
   if (noBackward) {
+    if (wait_duration.count() <= 0.0) {
+        TORCH_WARN("Leaking autograd threads");
+        return;
+    }
     for (auto& queue : device_ready_queues_) {
      queue->pushShutdownTask();
     }
     // Do not wait for termination of global threads on Windows
     // Because CRT terminates DLL threads before calling
     // global object destructors
 #if !defined(_WIN32) || defined(C10_USE_MSVC_STATIC_RUNTIME)
+
     std::unique_lock<std::mutex> lk(non_reentrant_device_thread_mutex_);
     while(non_reentrant_device_thread_count_.load() != 0) {
-      non_reentrant_device_thread_condvar_.wait(lk);
+      if (non_reentrant_device_thread_condvar_.wait_for(lk, wait_duration) == std::cv_status::timeout) {
+        TORCH_WARN("Leaking autograd threads");
```

Comment:
Not sure we want to warn here:
- TORCH_WARN the runtime might try to convert this to a python warning, which is not going to play nicely if python is being terminated
- It might spam some users that will get a warning every time they exit pytorch. Since we don't warn already if noBackward is false, I don't think we need to warn here.

Reply:
Agreed on the spamming. Also not sure the logic here is quite correct; if there are spurious wakeups on the `wait_for` you could still end up waiting indefinitely because you do the wait duration every time.

Reply:
I think the warning should stay, if timeout is non-zero, but should not warn if it is explicitly set to 0.
Will replace `wait_for` with `wait_until`

Reply:
Well as of today, if you quit while some autograd work is ongoing on another thread, you leak everything and there is no warrning.
While I understand that some power users would want to know, I think most users won't care. And I am not sure that raising a warning here is safe.
Maybe we want to make this a DEBUG build only warning and use `c10::Warning::warn()` to make sure you never call into unexpected CPython APIs.

Reply:
Looks like you have strong objection against the warnings here. Removed them. (Although again, when wait timeout is non-zero users should not see them by default)